Allow renaming of corruption loadouts, add 4th loadout

strentax opened the pull request:
Players requested more loadouts for corruptions due to complexity as game goes on. Adding one more for now until design decisions for further are discussed.

Here are the changes:
Corruptions.ts: Move loadout name label to separate function, add function to prompt for name.

Synergism.ts: Import new corruption loadout name function, call during other display updates. Add 4th loadout. Add new corruptionLoadoutNames type initialization.

Synergism.d.ts: add corruptionLoadoutNames Record declaration.

Synergism.css: add styling for corrLoadoutName class.

KhafraDev commented on src/Corruptions.ts:
```diff
@@ -232,6 +236,36 @@ const corruptionLoadoutSaveLoad = (save = true, loadout = 1) => {
 
+async function corruptionLoadoutGetNewName(loadout = 1) {
+    const maxChars = 9
+    let renamePrompt = await Prompt('What would you like to name Loadout ' + loadout + '? Names cannot be longer than ' + maxChars + ' characters. Nothing crazy!');
```
```suggestion
    const renamePrompt = await Prompt(
        `What would you like to name Loadout ${loadout}? ` +
        `Names cannot be longer than ${maxChars} characters. Nothing crazy!`
    );
```
so it doesn't go off the page

KhafraDev commented on src/Corruptions.ts:
```diff
@@ -232,6 +236,36 @@ const corruptionLoadoutSaveLoad = (save = true, loadout = 1) => {
+async function corruptionLoadoutGetNewName(loadout = 1) {
+    const maxChars = 9
+    let renamePrompt = await Prompt('What would you like to name Loadout ' + loadout + '? Names cannot be longer than ' + maxChars + ' characters. Nothing crazy!');
+    renamePrompt = String(renamePrompt)
```
```suggestion

```
String(null) === 'null' which breaks the null check below, plus it's unneeded

KhafraDev commented on src/Corruptions.ts:
```diff
@@ -232,6 +236,36 @@ const corruptionLoadoutSaveLoad = (save = true, loadout = 1) => {
+    const maxChars = 9
+    let renamePrompt = await Prompt('What would you like to name Loadout ' + loadout + '? Names cannot be longer than ' + maxChars + ' characters. Nothing crazy!');
+    renamePrompt = String(renamePrompt)
+    let regex = /^[\x00-\xFF]*$/
```
```suggestion
    const regex = /^[\x00-\xFF]*$/
```

KhafraDev commented on src/Corruptions.ts:
```diff
@@ -232,6 +236,36 @@ const corruptionLoadoutSaveLoad = (save = true, loadout = 1) => {
+    renamePrompt = String(renamePrompt)
+    let regex = /^[\x00-\xFF]*$/
+
+    if (!renamePrompt)
+        return Alert('Okay, maybe next time.');
+    else if (renamePrompt.length > maxChars)
+        return Alert('The name you provided is too long! Try again.')
+    else if (!renamePrompt.match(regex))
+        return Alert("The Loadout Renamer didn't like a character in your name! Try something else.")
+    else {
+        player.corruptionLoadoutNames[loadout] = renamePrompt
+        updateCorruptionLoadoutNames();
+    }
```
```suggestion
    if (!renamePrompt) {
        return Alert('Okay, maybe next time.');
    } else if (renamePrompt.length > maxChars) {
        return Alert('The name you provided is too long! Try again.')
    } else if (!renamePrompt.match(regex)) {
        return Alert("The Loadout Renamer didn't like a character in your name! Try something else.")
    } else {
        player.corruptionLoadoutNames[loadout] = renamePrompt
        updateCorruptionLoadoutNames();
    }
```

KhafraDev commented on src/Corruptions.ts:
```diff
@@ -232,6 +236,36 @@ const corruptionLoadoutSaveLoad = (save = true, loadout = 1) => {
+        return Alert('Okay, maybe next time.');
+    else if (renamePrompt.length > maxChars)
+        return Alert('The name you provided is too long! Try again.')
+    else if (!renamePrompt.match(regex))
```
```suggestion
    } else if (!regex.test(renamePrompt) {
```

KhafraDev commented on src/Corruptions.ts:
```diff
@@ -232,6 +236,36 @@ const corruptionLoadoutSaveLoad = (save = true, loadout = 1) => {
+export const updateCorruptionLoadoutNames = () => {
+    const rows = getElementById<HTMLTableElement>("corruptionLoadoutTable").rows
+    for (let i = 1; i < Object.keys(player.corruptionLoadouts).length + 1; i++) {
+        let cells = rows[i + 1].cells
```
```suggestion
        const cells = rows[i + 1].cells
```

KhafraDev commented on src/Corruptions.ts:
```diff
@@ -232,6 +236,36 @@ const corruptionLoadoutSaveLoad = (save = true, loadout = 1) => {
+    for (let i = 1; i < Object.keys(player.corruptionLoadouts).length + 1; i++) {
+        let cells = rows[i + 1].cells
+        if (!cells[0].innerHTML) {  //first time setup
+            cells[0].onclick = () => corruptionLoadoutGetNewName(i);
```
```suggestion
            cells[0].addEventListener('click', () => corruptionLoadoutGetNewName(i));
```

KhafraDev commented on src/Corruptions.ts:
```diff
@@ -232,6 +236,36 @@ const corruptionLoadoutSaveLoad = (save = true, loadout = 1) => {
+        let cells = rows[i + 1].cells
+        if (!cells[0].innerHTML) {  //first time setup
+            cells[0].onclick = () => corruptionLoadoutGetNewName(i);
+            cells[0].className = 'corrLoadoutName'
```
```suggestion
            cells[0].classList.add('corrLoadoutName');
```

KhafraDev commented on src/Corruptions.ts:
```diff
@@ -232,6 +236,36 @@ const corruptionLoadoutSaveLoad = (save = true, loadout = 1) => {
+            cells[0].onclick = () => corruptionLoadoutGetNewName(i);
+            cells[0].className = 'corrLoadoutName'
+        }
+        cells[0].innerHTML = player.corruptionLoadoutNames[i] + ':'
```
```suggestion
        cells[0].textContent = `${player.corruptionLoadoutNames[i]}:`;
```

KhafraDev commented on src/Corruptions.ts:
```diff
@@ -232,6 +236,36 @@ const corruptionLoadoutSaveLoad = (save = true, loadout = 1) => {
+    const rows = getElementById<HTMLTableElement>("corruptionLoadoutTable").rows
+    for (let i = 1; i < Object.keys(player.corruptionLoadouts).length + 1; i++) {
+        let cells = rows[i + 1].cells
+        if (!cells[0].innerHTML) {  //first time setup
```
```suggestion
        if (cells[0].textContent.length === 0) {  //first time setup
```
to be less ambiguous

KhafraDev commented on src/Synergism.ts:
```diff
@@ -572,10 +572,17 @@ export const player: Player = {
+        3: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+        4: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+    },
+    corruptionLoadoutNames: {
```
this should just be an array

KhafraDev commented on src/types/Synergism.d.ts:
```diff
@@ -448,6 +448,7 @@ export interface Player {
     prototypeCorruptions: number[]
     usedCorruptions: number[]
     corruptionLoadouts: Record<number, number[]>
+    corruptionLoadoutNames: Record<number, string>
```
same here

strentax commented:
@KhafraDev I addressed your requested changes. I also got annoyed by the name column width bouncing around, so I added a class to the <th> in index.html, and in the css just assigned it a relative width of 17%. Seems to work nice.

KhafraDev commented on src/Corruptions.ts:
```diff
@@ -232,6 +236,41 @@ const corruptionLoadoutSaveLoad = (save = true, loadout = 1) => {
     }
 }
 
+async function corruptionLoadoutGetNewName(loadout = 1) {
```
this should default to 0 now that it's an array

KhafraDev commented on src/Corruptions.ts:
```diff
@@ -232,6 +236,41 @@ const corruptionLoadoutSaveLoad = (save = true, loadout = 1) => {
+        return Alert("The Loadout Renamer didn't like a character in your name! Try something else.")
+    }
+    else {
+        player.corruptionLoadoutNames[loadout] = renamePrompt
```
or this needs to be loadout - 1

KhafraDev commented on src/Corruptions.ts:
```diff
@@ -232,6 +236,41 @@ const corruptionLoadoutSaveLoad = (save = true, loadout = 1) => {
+    for (let i = 1; i < Object.keys(player.corruptionLoadouts).length + 1; i++) {
+        const cells = rows[i + 1].cells
+        if (cells[0].textContent.length === 0) {  //first time setup
+            cells[0].addEventListener('click', () => corruptionLoadoutGetNewName(i));
```
i - 1

KhafraDev commented on src/Corruptions.ts:
```diff
@@ -232,6 +236,41 @@ const corruptionLoadoutSaveLoad = (save = true, loadout = 1) => {
+            cells[0].addEventListener('click', () => corruptionLoadoutGetNewName(i));
+            cells[0].classList.add('corrLoadoutName');
+        }
+        cells[0].textContent = `${player.corruptionLoadoutNames[i]}:`;
```
i - 1

strentax commented:
@KhafraDev addressed requests, plus noticed a double - 1 adjustment so fixed that

KhafraDev commented on src/Corruptions.ts:
```diff
@@ -232,6 +236,41 @@ const corruptionLoadoutSaveLoad = (save = true, loadout = 1) => {
+
+export const updateCorruptionLoadoutNames = () => {
+    const rows = getElementById<HTMLTableElement>("corruptionLoadoutTable").rows
+    for (let i = 1; i < Object.keys(player.corruptionLoadouts).length + 1; i++) {
+        const cells = rows[i + 1].cells
```
```suggestion
    for (let i = 0; i < Object.keys(player.corruptionLoadouts).length + 1; i++) {
        const cells = rows[i + 2].cells
```
this fixes the issues

KhafraDev commented on src/Corruptions.ts:
```diff
@@ -232,6 +236,41 @@ const corruptionLoadoutSaveLoad = (save = true, loadout = 1) => {
+            cells[0].addEventListener('click', () => corruptionLoadoutGetNewName(i)); //get name function handles -1 for array
+            cells[0].classList.add('corrLoadoutName');
+        }
+        cells[0].textContent = `${player.corruptionLoadoutNames[i - 1]}:`;
```
```suggestion
        cells[0].textContent = `${player.corruptionLoadoutNames[i]}:`;
```
with previous comment

strentax commented:
Updated and fixed exception

KhafraDev commented:
doesn't fix the issue

strentax commented:
When you run that code with your changes, you don't get an exception? It's sticking at the offline progression screen because of the error. It stops when I take the length + 1 out.

strentax commented:
`Object.keys(player.corruptionLoadouts).length + 1` = 5
i < 5 is the loop range
i = 4 will cause player.corruptionLoadoutNames[i] to fail
